src/ghcup/hls.rs

```diff
@@ -0,0 +1,72 @@
+use async_trait::async_trait;
+use slog::info;
+use structopt::StructOpt;
+
+use crate::common::{Mission, SnapshotConfig, TransferURL};
+use crate::error::Result;
+use crate::metadata::SnapshotMeta;
+use crate::traits::{SnapshotStorage, SourceStorage};
+
+use super::parser::GhcupYamlParser;
+use super::utils::get_yaml_url;
+
+#[derive(Debug, Clone, StructOpt)]
+pub struct GhcupHLS {
+    #[structopt(
+        long,
+        help = "Ghcup upstream",
+        default_value = "https://gitlab.haskell.org/haskell/ghcup-hs/-/raw/master/"
+    )]
+    pub ghcup_base: String,
+}
+
+#[async_trait]
+impl SnapshotStorage<SnapshotMeta> for GhcupHLS {
+    async fn snapshot(
+        &mut self,
+        mission: Mission,
+        _config: &SnapshotConfig,
+    ) -> Result<Vec<SnapshotMeta>> {
+        let logger = mission.logger;
+        let progress = mission.progress;
+        let client = mission.client;
+
+        let base_url = self.ghcup_base.trim_end_matches('/');
+
+        info!(logger, "fetching ghcup config...");
+        progress.set_message("downloading version file");
+        let yaml_url = get_yaml_url(base_url, &client).await?;
+        progress.set_message("downloading yaml config");
+        let yaml_data = client.get(&yaml_url).send().await?.bytes().await?;
+        let ghcup_config: GhcupYamlParser = serde_yaml::from_slice(&yaml_data)?;
+
+        let fetch_uris: Vec<_> = ghcup_config
+            .ghcup_downloads
+            .uris(true)
+            .into_iter()
+            .filter_map(|s| {
+                s.strip_prefix(
```

Comment:
I would suggest panic if there's prefix mismatch.

Reply:
It's expected to have prefix mismatch because there will be packages starts with 'kernel.org', not 'github.com'.
Maybe we should rename 'GhcupPackages' and 'GhcupHLS' as 'GhcupHaskell' and 'GhcupGithub'?